drivers/usb/udc/udc_stm32.c

```diff
@@ -0,0 +1,1128 @@
+
+	LOG_DBG("");
+
+	udc_set_suspended(priv->dev, true);
+	udc_submit_event(priv->dev, UDC_EVT_SUSPEND, 0, NULL);
+}
+
+void HAL_PCD_ResumeCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+
+	LOG_DBG("");
+
+	udc_set_suspended(priv->dev, false);
+	udc_submit_event(priv->dev, UDC_EVT_RESUME, 0, NULL);
+}
+
+static int usbd_ctrl_feed_dout(const struct device *dev, const size_t length)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	struct udc_ep_config *cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
+	struct net_buf *buf;
+
+	LOG_DBG("");
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, length);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	net_buf_put(&cfg->fifo, buf);
+
+	HAL_PCD_EP_Receive(&priv->pcd, cfg->addr, buf->data, buf->size);
+
+	return 0;
+}
+
+void HAL_PCD_SetupStageCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+	struct usb_setup_packet *setup = (void *)priv->pcd.Setup;
+	const struct device *dev = priv->dev;
+	struct net_buf *buf;
+	int err;
+
+	LOG_DBG("");
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT,
+			     sizeof(struct usb_setup_packet));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate for setup");
+		return;
+	}
+
+	udc_ep_buf_set_setup(buf);
+	memcpy(buf->data, setup, 8);
+	net_buf_add(buf, 8);
+
+	udc_ctrl_update_stage(dev, buf);
+
+	if (!buf->len) {
+		return;
+	}
+
+	if (udc_ctrl_stage_is_data_out(dev)) {
+		/*  Allocate and feed buffer for data OUT stage */
+		err = usbd_ctrl_feed_dout(dev, udc_data_stage_length(buf));
+		if (err == -ENOMEM) {
+			udc_submit_event(dev, UDC_EVT_EP_REQUEST, err, buf);
+		}
+	} else if (udc_ctrl_stage_is_data_in(dev)) {
+		udc_ctrl_submit_s_in_status(dev);
+	} else {
+		udc_ctrl_submit_s_status(dev);
+	}
+
+	if (setup->bRequest == USB_SREQ_SET_ADDRESS) {
+		HAL_PCD_SetAddress(&priv->pcd, setup->wValue);
+	}
+}
+
+void HAL_PCD_SOFCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+
+	udc_submit_event(priv->dev, UDC_EVT_SOF, 0, NULL);
+}
+
+static int udc_stm32_tx(const struct device *dev, uint8_t ep,
+			struct net_buf *buf)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	const struct udc_stm32_config *cfg = dev->config;
+	uint8_t *data; uint32_t len;
+	HAL_StatusTypeDef status;
+
+	LOG_DBG("0x%02x len %u", ep, buf->len);
+
+	if (udc_ep_is_busy(dev, ep)) {
+		return 0;
+	}
+
+	data = buf->data;
+	len = buf->len;
+
+	if (ep == USB_CONTROL_EP_IN) {
+		len = MIN(cfg->ep0_mps, buf->len);
+	}
+
+	buf->data += len;
+	buf->len -= len;
+
+	status = HAL_PCD_EP_Transmit(&priv->pcd, ep, data, len);
+	if (status != HAL_OK) {
+		LOG_ERR("HAL_PCD_EP_Transmit failed(0x%02x), %d", ep, (int)status);
+		return -EIO;
+	}
+
+	udc_ep_set_busy(dev, ep, true);
+
+	if (ep == USB_CONTROL_EP_IN && len > 0) {
+		/* Wait for an empty package from the host.
+		 * This also flushes the TX FIFO to the host.
+		 */
+		usbd_ctrl_feed_dout(dev, 0);
```

Comment:
I don't understand what this does. How can there be "empty package" from the host on IN endpoint (IN is always from device to host)?! My only suspicion is that it arms the STATUS stage in a obfuscated way.

Reply:
> How can there be "empty package" from the host on IN endpoint

It's OUT endpoint here, if we submit IN data stage to the Host, we also need to have OUT ready for the status stage.

> My only suspicion is that it arms the STATUS stage in a obfuscated way.

Seems the data stage is not flushed to the bus until we have status allocated...